src/sqlfluff/dialects/dialect_mysql.py

```diff
@@ -270,11 +410,374 @@ class DropStatementSegment(BaseSegment):
+        Ref("SingleIdentifierGrammar"),
+    )
+
+
+@mysql_dialect.segment(replace=True)
+class SelectClauseModifierSegment(BaseSegment):
+    """Things that come after SELECT but before the columns."""
+
+    type = "select_clause_modifier"
+    match_grammar = Sequence(
+        OneOf("DISTINCT", "ALL", "DISTINCTROW", optional=True),
+        Ref.keyword("HIGH_PRIORITY", optional=True),
+        Ref.keyword("STRAIGHT_JOIN", optional=True),
+        Ref.keyword("SQL_SMALL_RESULT", optional=True),
+        Ref.keyword("SQL_BIG_RESULT", optional=True),
+        Ref.keyword("SQL_BUFFER_RESULT", optional=True),
+        Ref.keyword("SQL_CACHE", optional=True),
+        Ref.keyword("SQL_NO_CACHE", optional=True),
+        Ref.keyword("SQL_CALC_FOUND_ROWS", optional=True),
+        optional=True,
+    )
+
+
+@mysql_dialect.segment()
+class IntoClauseSegment(BaseSegment):
+    """This is an `INTO` clause for assigning variables in a select statement.
+
+    https://dev.mysql.com/doc/refman/5.7/en/load-data.html
+    https://dev.mysql.com/doc/refman/5.7/en/select-into.html
+    """
+
+    type = "into_clause"
+
+    match_grammar = Sequence(
+        "INTO",
+        OneOf(
+            Delimited(
+                AnyNumberOf(
+                    Ref("SessionVariableNameSegment"),
+                    Ref("LocalVariableNameSegment"),
+                ),
+            ),
+            Sequence("DUMPFILE", Ref("QuotedLiteralSegment")),
+            Sequence(
+                "OUTFILE",
+                Ref("QuotedLiteralSegment"),
+                Sequence(
+                    "CHARACTER", "SET", Ref("NakedIdentifierSegment"), optional=True
+                ),
+                Sequence(
+                    OneOf("FIELDS", "COLUMNS"),
+                    Sequence(
+                        "TERMINATED", "BY", Ref("QuotedLiteralSegment"), optional=True
+                    ),
+                    Sequence(
+                        Ref.keyword("OPTIONALLY", optional=True),
+                        "ENCLOSED",
+                        "BY",
+                        Ref("QuotedLiteralSegment"),
+                        optional=True,
+                    ),
+                    Sequence(
+                        "ESCAPED", "BY", Ref("QuotedLiteralSegment"), optional=True
+                    ),
+                    optional=True,
+                ),
+                Sequence(
+                    "LINES",
+                    Sequence(
+                        "STARTING", "BY", Ref("QuotedLiteralSegment"), optional=True
+                    ),
+                    Sequence(
+                        "TERMINATED", "BY", Ref("QuotedLiteralSegment"), optional=True
+                    ),
+                    optional=True,
+                ),
+            ),
+        ),
+    )
+
+
+# I wanted to do an override and just insert the new Refs but I need
+# to have the IntoClause be before FromClauseSegment, but it doesn't work to add the `before` clause
```

Comment:
Reading into this more - can you change the PR to be the way you're expecting this to work so that we can debug more easily? As in:

```
    parse_grammar = ansi_dialect.get_segment(
        "UnorderedSelectStatementSegment"
    ).parse_grammar.copy(
        insert=[Ref("IntoClauseSegment", optional=True)],
        before=Ref("FromClauseSegment", optional=True),
    )
```

Reply:
I added the statements as I think it should work, but then it stops parsing correctly.  Here's an example of something that did parse, but now doesn't after this change.
[select into outfile](https://github.com/silverbullettruck/sqlfluff/blob/AddSelectStmtSegments/test/fixtures/parser/mysql/select_into_outfile_fields_enclosed.yml)